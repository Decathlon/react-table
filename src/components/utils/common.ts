const objectProto = Object.prototype;
const { hasOwnProperty } = objectProto;

export const isEmptyObj = (value: object) => {
  for (const key in value) {
    if (hasOwnProperty.call(value, key)) {
      return false;
    }
  }
  return true;
};
