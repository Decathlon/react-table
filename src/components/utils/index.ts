export * from "./table";
export * from "./table-selection";
export * from "./common";

export { default as shallowEqual } from "./shallowEqual";
