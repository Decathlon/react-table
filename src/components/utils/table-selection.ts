import { ISelectedCells } from "../table-selection/selection-handler";
import { ICellCoordinates, ICell } from "../table/cell";

export const isVerticalSelection = (selectedCells: ISelectedCells) => {
  const rowIds = Object.keys(selectedCells);
  return rowIds.length >= 1 && rowIds.some(rowId => selectedCells[rowId].length === 1);
};

export const isHorizontalSelection = (selectedCells: ISelectedCells) => {
  const rowIds = Object.keys(selectedCells);
  return rowIds.length === 1 && rowIds.some(rowId => selectedCells[rowId].length === 1);
};

export const isMultiDimensionSelection = (selectedCells: ISelectedCells) => {
  const rowIds = Object.keys(selectedCells);
  return rowIds.length > 1 && rowIds.some(rowId => selectedCells[rowId].length > 1);
};

export function getSelectedCellsProps<IDataCoordinates = any>(
  selectedCells: ISelectedCells,
  getCell: (cellCoordinates: ICellCoordinates) => ICell<IDataCoordinates>
) {
  return Object.keys(selectedCells).reduce<ICell<IDataCoordinates>[]>((result, rowIndex) => {
    result.push(...selectedCells[rowIndex].map(cellIndex => getCell({ rowIndex: parseInt(rowIndex), cellIndex })));
    return result;
  }, []);
}
