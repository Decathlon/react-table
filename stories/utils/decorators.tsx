import { ThemeProvider, StyledEngineProvider, createTheme, adaptV4Theme } from "@mui/material/styles";

export const blueDkt = "#0082c3";

const muiDktTheme = createTheme(
  adaptV4Theme({
    typography: {
      fontFamily: "Roboto Condensed",
    },
    palette: {
      primary: {
        light: blueDkt,
        dark: blueDkt,
        main: blueDkt,
      },
      secondary: {
        light: blueDkt,
        dark: blueDkt,
        main: blueDkt,
      },
    },
  })
);

/**
 * The MUI theme provider wrapper
 *
 * @param {() => JSX.Element} attributes a function to render the child component.
 *
 * @return {JSX.Element} The child wrapped by the MUI theme provider.
 */
export function withThemeProvider(renderChild: () => JSX.Element): JSX.Element {
  return (
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={muiDktTheme}>{renderChild()}</ThemeProvider>
    </StyledEngineProvider>
  );
}
