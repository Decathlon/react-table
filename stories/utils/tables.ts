import { IRow } from "../../src/components/table/row";

export const simpleTable = ({
  customCell,
  customCellProps,
  loading = false,
}: {
  customCell?: React.ComponentType<any>;
  customCellProps?: object;
  loading?: boolean;
}): IRow[] => [
  {
    id: "front",
    isHeader: true,
    cells: [
      {
        id: "wawoo",
        value: "Wawooo!",
      },
      {
        id: "wah",
        value: "Wah!",
      },
      {
        id: "cool",
        value: "Cool!",
      },
    ],
  },
  {
    id: "front1",
    cells: [
      {
        id: "react",
        value: "React",
        cellContent: customCell,
        cellContentProps: customCellProps,
      },
      {
        id: "angular",
        value: "Angular",
      },
      {
        id: "vuejs",
        value: "VueJS",
      },
    ],
  },
  {
    id: "back",
    loading,
    cells: [
      {
        id: "python",
        value: "Python",
      },
      {
        id: "Go",
        value: "Go",
      },
      {
        id: "Java",
        value: "java",
      },
    ],
  },
  {
    id: "back1",
    cells: [
      {
        id: "python",
        value: "Python1",
      },
      {
        id: "Go",
        value: "Go",
      },
      {
        id: "Java",
        value: "java",
      },
    ],
  },
  {
    id: "back2",
    cells: [
      {
        id: "python",
        value: "Python2",
      },
      {
        id: "Go",
        value: "Go",
      },
      {
        id: "Java",
        value: "java",
      },
    ],
  },
  {
    id: "back3",
    cells: [
      {
        id: "python",
        value: "Python3",
      },
      {
        id: "Go",
        value: "Go",
      },
      {
        id: "Java",
        value: "java",
      },
    ],
  },
];

export const subRows = ({ subsubRows = [] }: { subsubRows?: IRow[] }): IRow[] => [
  {
    id: "foo",
    cells: [
      {
        id: "storybook",
        value: "Storybook",
      },
      {
        id: "jest",
        value: "Jest",
      },
      {
        id: "Enzym",
        value: "enzyme",
        subItems: subsubRows,
      },
    ],
  },
  {
    id: "back",
    cells: [
      {
        id: "docker",
        value: "Docker",
      },
      {
        id: "doccomp",
        value: "Docker compose",
      },
      {
        id: "Sentry",
        value: "Sentry",
        subItems: subsubRows,
      },
    ],
  },
];

export const subMiam: IRow[] = [
  {
    id: "miam",
    cells: [
      {
        id: "sushi",
        value: "Sushi",
      },
      {
        id: "pizza",
        value: "Pizza",
      },
      {
        id: "mafe",
        value: "Mafe",
      },
    ],
  },
];

export const tableWithSubItems = ({
  firstSubRows = [],
  secondSubRows = [],
}: {
  firstSubRows?: IRow[];
  secondSubRows?: IRow[];
}): IRow[] => {
  const table = [
    {
      id: "front",
      isHeader: true,
      isSelectable: false,
      cells: [
        {
          id: "wawoo",
          value: "Wawooo!",
        },
        {
          id: "wah",
          value: "Wah!",
        },
        {
          id: "cool",
          value: "Cool!",
        },
      ],
    },
    {
      title: "foo",
      id: "front1",
      cells: [
        {
          id: "react",
          value: "React",
          subItems: firstSubRows,
        },
        {
          id: "angular",
          value: "Angular",
        },
        {
          id: "vuejs",
          value: "VueJS",
        },
      ],
    },
    {
      id: "back",
      cells: [
        {
          id: "python",
          value: "Python",
        },
        {
          id: "Go",
          value: "Go",
          subItems: secondSubRows,
        },
        {
          id: "Java",
          value: "java",
        },
      ],
    },
  ];
  return table;
};

export const tableWithDifferentRowSizes: IRow[] = [
  {
    id: "front",
    isHeader: true,
    cells: [
      {
        id: "wawoo",
        value: "Wawooo!",
      },
      {
        id: "wah",
        value: "Wah!",
      },
      {
        id: "cool",
        value: "Cool!",
      },
    ],
  },
  {
    id: "front1",
    isHeader: false,
    cells: [
      {
        id: "foo",
        value: "foo",
      },
      {
        id: "bar",
        value: "bar",
      },
      {
        id: "fuzz",
        value: "fuzz",
      },
    ],
  },
  {
    id: "small",
    size: 25,
    cells: [
      {
        id: "small",
        value: "Wawooo!",
      },
      {
        id: "wah",
        value: "Wah!",
      },
      {
        id: "cool",
        value: "Cool!",
      },
    ],
  },
  {
    id: "medium",
    size: 40,
    cells: [
      {
        id: "medium",
        value: "React",
      },
      {
        id: "angular",
        value: "Angular",
      },
      {
        id: "vuejs",
        value: "VueJS",
      },
    ],
  },
  {
    id: "large",
    size: 80,
    cells: [
      {
        id: "large",
        value: "Python",
      },
      {
        id: "Go",
        value: "Go",
      },
      {
        id: "Java",
        value: "java",
      },
    ],
  },
];

export const tableWithDifferentColspan: IRow[] = [
  {
    id: "front",
    isHeader: true,
    cells: [
      {
        id: "wawoo",
        value: "Wawooo!",
      },
      {
        id: "wah",
        value: "Wah!",
      },
      {
        id: "cool",
        value: "Cool!",
      },
    ],
  },
  {
    id: "front1",
    isHeader: false,
    cells: [
      {
        id: "foo",
        value: "foo",
        colspan: 2,
      },
      {
        id: "fuzz",
        value: "fuzz",
      },
    ],
  },
  {
    id: "foo",
    cells: [
      {
        id: "wah",
        value: "Wah!",
      },
      {
        id: "cool",
        value: "Cool!",
        colspan: 2,
      },
    ],
  },
  {
    id: "foo",
    cells: [
      {
        id: "medium",
        value: "React",
      },
      {
        id: "angular",
        value: "Angular",
      },
      {
        id: "vuejs",
        value: "VueJS",
      },
    ],
  },
  {
    id: "foo",
    cells: [
      {
        id: "large",
        value: "Python",
        colspan: 3,
      },
    ],
  },
];

export const defaultColspanMatrix = (cellsCount: number) => [
  {
    cells: [
      {
        column: 0,
        value: "Table title",
        colspan: cellsCount,
        colspanMatrix: [
          {
            cells: [
              {
                column: 0,
                value: "****** My chart ******",
                colspan: cellsCount,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    cells: [
      {
        column: 1,
        colspan: 2,
      },
    ],
  },
];

export function generateTableWithCustomColspan(
  rowsCount: number,
  cellsCount: number,
  subRow = false,
  level = 0,
  colspanMatrix: any = defaultColspanMatrix(cellsCount)
) {
  const rows: IRow[] = Array.from(Array(rowsCount), (_, rowIndex) => {
    const totalColspan = colspanMatrix[rowIndex]
      ? colspanMatrix[rowIndex].cells.reduce((sum: number, currentValue: any) => {
          return sum + currentValue.colspan;
        }, 0)
      : 0;
    const cellsCountWithColspan =
      cellsCount - totalColspan + (colspanMatrix[rowIndex] ? colspanMatrix[rowIndex].cells.length : 0);
    const cells = generateRowWithCustomColumns(rowIndex, cellsCount, cellsCountWithColspan, subRow, level, colspanMatrix);

    return {
      id: `${rowIndex}`,
      isHeader: false,
      cells,
    };
  });

  return {
    id: `test`,
    rows,
  };
}

function generateRowWithCustomColumns(
  rowIndex: number,
  cellsCount: number,
  cellsCountWithColspan: number,
  subRow: boolean,
  level: number,
  colspanMatrix: any
) {
  return Array.from(Array(cellsCountWithColspan), (_, cellIndex) => {
    const currentCell = colspanMatrix[rowIndex]
      ? colspanMatrix[rowIndex].cells.find((cell: any) => cell.column === cellIndex && cell.colspan)
      : null;
    if (currentCell && currentCell.colspan >= 1) {
      return {
        id: `(${rowIndex},${cellIndex})-${level}`,
        value: currentCell.value || `(${rowIndex},${cellIndex})`,
        colspan: currentCell.colspan,
        subItems:
          subRow && cellIndex === 0
            ? Array.from(Array(cellsCount), (_, rowIndex) =>
                generateRow(rowIndex, cellsCount, level <= 2, level + 1, currentCell.colspanMatrix)
              )
            : [],
      };
    }
    return {
      id: `(${rowIndex},${cellIndex})-${level}`,
      value: `(${rowIndex},${cellIndex})`,
    };
  });
}

export function generateRow(index: number, cellsCount: number, subRow = false, level = 0, colspanMatrix?: any): IRow {
  const cells = Array.from(Array(cellsCount), (_, cellIndex) => {
    const currentCell =
      colspanMatrix && colspanMatrix[index]
        ? colspanMatrix[index].cells.find((cell: any) => cell.column === cellIndex && cell.colspan)
        : null;
    return {
      id: `(${index},${cellIndex})-${level}`,
      value: (currentCell && currentCell.value) || `(${index},${cellIndex})`,
      colspan: currentCell ? currentCell.colspan : 1,
      subItems:
        subRow && cellIndex === 0 && index === 2
          ? Array.from(Array(20), (_, rowIndex) => generateRow(rowIndex, cellsCount, level <= 2, level + 1))
          : [],
    };
  });
  return {
    id: `${index}`,
    isHeader: index === 0,
    cells,
  };
}

export function generateTable(rowsCount: number, cellsCount: number, props = {}, subRow = false) {
  const rows = Array.from(Array(rowsCount), (_, rowIndex) => generateRow(rowIndex, cellsCount, subRow));
  return {
    id: "table-foo",
    ...props,
    rows,
  };
}
