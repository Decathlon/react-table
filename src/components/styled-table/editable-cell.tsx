import * as React from "react";
import TextField from "@material-ui/core/TextField";
import * as NumberFormat from "react-number-format";
import classnames from "classnames";

import { IContentCellProps } from "../table/cell";
import { Nullable } from "../typing";
import { getStringNumberWithoutTrailingZeros } from "../utils";

// @ts-ignore https://github.com/s-yadav/react-number-format/issues/180
const NumberFormatComponent = NumberFormat.default || NumberFormat;

export const EDITED_CELL_CLASSNAME = "edited-cell";
export const KEYCODE_ENTER = 13;
const minInputWidth = 20; // Equal to the minimum editable cell underline width

export interface IMask {
  decimals: number;
  is_percentage: boolean;
  is_negative: boolean;
}

/**
 * initial_value is required to differentiate
 * the case when input is set to value "0"
 * and the case when input isn't setted yet and the value is "null"
 *
 * If input is cleared, value is set to initial_value
 */
export interface IEdiTableCellProps extends IContentCellProps {
  isEdited: boolean;
  initial_value: Nullable<number>;
  value: Nullable<number>;
  mask: IMask;
  formatValue: (value: Nullable<number>, mask?: IMask) => string;
  onConfirmValue: (value: Nullable<number>) => void;
  validateValue?: (value: Nullable<number>) => boolean;
}

interface IState {
  inputValue: string;
  /** determine if the cell is focused */
  isFocused: boolean;
  /** determine if new input value is valid (check with validateValue method) and can be apply */
  isValidValue: boolean;
}

export default class EdiTableCell extends React.PureComponent<IEdiTableCellProps, IState> {
  constructor(props: IEdiTableCellProps) {
    super(props);
    const { value } = this.props;
    const newInputValue = this.getInputValue();
    this.state = {
      inputValue: newInputValue,
      isFocused: false,
      isValidValue: this.isValidValue(value)
    };
  }

  public componentWillUnmount() {
    const { isFocused } = this.state;
    if (isFocused) {
      const { onConfirmValue } = this.props;
      const [hasChanged, newValue] = this.getNewValue();
      if (hasChanged) {
        onConfirmValue(newValue);
      }
    }
  }

  private getNewValue = (): [boolean, Nullable<number>] => {
    const { inputValue } = this.state;
    const { initial_value, value, mask } = this.props;
    let newValue: Nullable<number> = parseFloat(inputValue);
    newValue = isNaN(newValue) ? null : newValue;
    /**
     * Change percentage value format because :
     * Edited value has the following format : 100 is equal to 100%
     * Data we worked on has the following format : 1 is equal to 100
     */
    if (mask.is_percentage) {
      newValue = newValue ? newValue / 100 : newValue;
    }
    /** if the edited value is different than the value before edition */
    if (newValue !== value) {
      const isCleared = inputValue === "";
      if (isCleared) {
        newValue = initial_value === null ? null : 0;
      }
      return [true, newValue];
    }
    return [false, newValue];
  };

  private focus = () => {
    const { value } = this.props;
    const newInputValue = this.getInputValue();
    this.setState({
      isFocused: true,
      inputValue: newInputValue,
      isValidValue: this.isValidValue(value)
    });
  };

  private getInputValue = () => {
    const { value, mask } = this.props;
    let newInputValue = value;
    const isNumber = value !== null && !isNaN(value);
    if (isNumber && mask.is_percentage) {
      // @ts-ignore value !== null
      newInputValue = value * 100;
    }
    // @ts-ignore value !== null
    return isNumber ? getStringNumberWithoutTrailingZeros(newInputValue, mask.decimals) : "";
  };

  protected clearFocus = () => {
    this.setState({
      isFocused: false
    });
  };

  private isValidValue = (value: Nullable<number>) => {
    const { validateValue } = this.props;
    const isNumber = value !== null && !isNaN(value);
    return validateValue && isNumber ? validateValue(value) : true;
  };

  private isAllowed = (values: NumberFormat.NumberFormatValues) => {
    const { isValidValue: currentIsValidValue } = this.state;
    const isValidValue = this.isValidValue(values.floatValue);

    if (currentIsValidValue !== isValidValue) {
      this.setState({
        isValidValue
      });
    }
    return isValidValue;
  };

  private onValueChangeHandle = (values: NumberFormat.NumberFormatValues) => {
    this.setState({
      inputValue: values.value
    });
  };

  private onKeyPress = (event: React.KeyboardEvent): void => {
    if (event.which === KEYCODE_ENTER || event.keyCode === KEYCODE_ENTER) {
      this.onBlur();
    }
  };

  private onBlur = (): void => {
    const { onConfirmValue } = this.props;
    const [hasChanged, newValue] = this.getNewValue();
    if (hasChanged) {
      this.setState({ inputValue: "" }, () => {
        onConfirmValue(newValue);
      });
    }
    this.clearFocus();
  };

  public render() {
    const { inputValue, isFocused, isValidValue } = this.state;
    const { isEdited, value, mask, formatValue } = this.props;
    let inputValueWidth = inputValue !== "" ? inputValue.trim().length * 10 : minInputWidth;
    inputValueWidth = inputValueWidth < minInputWidth ? minInputWidth : inputValueWidth;
    return (
      <div
        className={classnames("editable-cell", {
          empty: !isFocused && value === null,
          error: !isValidValue
        })}
        data-testid="editable-cell"
        onClick={!isFocused ? this.focus : undefined}
      >
        {isFocused ? (
          // @ts-ignore https://github.com/s-yadav/react-number-format/issues/180
          <NumberFormatComponent
            autoFocus
            data-testid="editable-cell-text-field"
            customInput={TextField}
            defaultValue={inputValue}
            onValueChange={this.onValueChangeHandle}
            onBlur={this.onBlur}
            onKeyPress={this.onKeyPress}
            thousandSeparator=" "
            decimalSeparator=","
            style={{ width: inputValueWidth }}
            InputProps={{ classes: { underline: "editable-cell__underline" } }}
            isNumericString
            decimalScale={mask.decimals}
            allowNegative={mask.is_negative}
            isAllowed={this.isAllowed}
          />
        ) : (
          <div
            className={classnames("editable-cell__value", {
              [EDITED_CELL_CLASSNAME]: isEdited
            })}
          >
            <span className="text">{formatValue(value, mask)}</span>
          </div>
        )}
      </div>
    );
  }
}
