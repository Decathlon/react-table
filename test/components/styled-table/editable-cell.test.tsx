/// <reference path="../../typings/tests-entry.d.ts" />

import * as React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { mount } from "enzyme";

import EdiTableCell, { KEYCODE_ENTER, IEdiTableCellProps } from "../../../src/components/styled-table/editable-cell";
import { withThemeProvider } from "../../../stories/utils/decorators";
import { formatValue } from "../../../stories/components/styled-table/editable-cell.stories";

describe("EdiTableCell component", () => {
  const mask = {
    style: "currency",
    currency: "EUR",
    decimals: 2,
    is_percentage: false,
    is_negative: false
  };

  test("should render the default editable cell", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 0,
      initial_value: 0,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<EdiTableCell {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render editable cell with null value : '-'", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: null,
      initial_value: null,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<EdiTableCell {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render editable cell with percentage value : '100.00 %'", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 1,
      initial_value: 1,
      onConfirmValue: jest.fn(),
      mask: { ...mask, is_percentage: true },
      formatValue
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<EdiTableCell {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should render edited cell", () => {
    const props: IEdiTableCellProps = {
      isEdited: true,
      value: 50,
      initial_value: null,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const shallowRenderer = createRenderer();
    shallowRenderer.render(<EdiTableCell {...props} />);
    const rendered = shallowRenderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  test("should set the focus on cell click", () => {
    const props: IEdiTableCellProps = {
      isEdited: true,
      value: 0,
      initial_value: 0,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // onClick
    wrapper.find("[data-testid='editable-cell']").simulate("click");
    expect(editableCellInstance.state.isFocused).toBeTruthy();
  });

  test("should call onConfirmValue on blur with simple value", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 0,
      initial_value: 0,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "22" });
    expect(editableCellInstance.state.inputValue).toEqual("22");
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: KEYCODE_ENTER });
    expect(props.onConfirmValue).toHaveBeenCalledTimes(1);
    expect(props.onConfirmValue).toHaveBeenCalledWith(22);
  });

  test("should call onConfirmValue on blur with percentage value", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 0,
      initial_value: 0,
      onConfirmValue: jest.fn(),
      mask: { ...mask, is_percentage: true },
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "22" });
    expect(editableCellInstance.state.inputValue).toEqual("22");
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: KEYCODE_ENTER });
    expect(props.onConfirmValue).toHaveBeenCalledTimes(1);
    expect(props.onConfirmValue).toHaveBeenCalledWith(0.22);
  });

  test("should call onConfirmValue on blur with null", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 22,
      initial_value: null,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "" });
    expect(editableCellInstance.state.inputValue).toEqual("");
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: KEYCODE_ENTER });
    expect(props.onConfirmValue).toHaveBeenCalledTimes(1);
    expect(props.onConfirmValue).toHaveBeenCalledWith(null);
  });

  test("should call onConfirmValue on blur with 0", () => {
    const props: IEdiTableCellProps = {
      isEdited: false,
      value: 22,
      initial_value: 22,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "" });
    expect(editableCellInstance.state.inputValue).toEqual("");
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: KEYCODE_ENTER });
    expect(props.onConfirmValue).toHaveBeenCalledTimes(1);
    expect(props.onConfirmValue).toHaveBeenCalledWith(0);
  });

  test("shouldn't call onConfirmValue on blur when input value has not changed", () => {
    const props: IEdiTableCellProps = {
      isEdited: true,
      value: 22,
      initial_value: 30,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "22" });
    expect(editableCellInstance.state.inputValue).toEqual("22");
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: KEYCODE_ENTER });
    expect(props.onConfirmValue).toHaveBeenCalledTimes(0);
  });

  test("shouldn't call onConfirmValue with wrong keyPress", () => {
    const props: IEdiTableCellProps = {
      isEdited: true,
      value: 0,
      initial_value: 0,
      onConfirmValue: jest.fn(),
      mask,
      formatValue
    };
    const wrapper = mount(withThemeProvider(() => <EdiTableCell {...props} />));
    const editableCellInstance: EdiTableCell = wrapper.find(EdiTableCell).instance() as EdiTableCell;
    // @ts-ignore onValueChangeHandle is a private method
    editableCellInstance.onValueChangeHandle({ value: "22" });
    // @ts-ignore onKeyPress is a private method and is not assignable to parameter of type 'KeyBoardEvent'
    editableCellInstance.onKeyPress({ keyCode: 14 });
    expect(props.onConfirmValue).toBeCalledTimes(0);
  });
});
