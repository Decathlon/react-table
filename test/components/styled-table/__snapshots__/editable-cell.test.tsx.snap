// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdiTableCell component should render a disabled editable cell 1`] = `
<div
  className="editable-cell"
  data-testid="editable-cell"
>
  <div
    className="editable-cell__value"
    title="0.00"
  >
    <span
      className="text"
    >
      0.00
    </span>
  </div>
</div>
`;

exports[`EdiTableCell component should render editable cell with null value : '-' 1`] = `
<div
  className="editable-cell empty"
  data-testid="editable-cell"
  onClick={[Function]}
>
  <div
    className="editable-cell__value"
    title="-"
  >
    <span
      className="text"
    >
      -
    </span>
  </div>
</div>
`;

exports[`EdiTableCell component should render editable cell with percentage value : '100.00 %' 1`] = `
<div
  className="editable-cell"
  data-testid="editable-cell"
  onClick={[Function]}
>
  <div
    className="editable-cell__value"
    title="100.00%"
  >
    <span
      className="text"
    >
      100.00%
    </span>
  </div>
</div>
`;

exports[`EdiTableCell component should render edited cell 1`] = `
<div
  className="editable-cell"
  data-testid="editable-cell"
  onClick={[Function]}
>
  <div
    className="editable-cell__value edited-cell"
    title="50.00"
  >
    <span
      className="text"
    >
      50.00
    </span>
  </div>
</div>
`;

exports[`EdiTableCell component should render the default editable cell 1`] = `
<div
  className="editable-cell"
  data-testid="editable-cell"
  onClick={[Function]}
>
  <div
    className="editable-cell__value"
    title="0.00"
  >
    <span
      className="text"
    >
      0.00
    </span>
  </div>
</div>
`;
